### PR TITLE
Dev/initial

### PR DESCRIPTION
**Pull request by janstaelensskyline**

### CODE REVIEW

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+			_logger.LogInformation(JsonConvert.SerializeObject(result));
+			return result;
+		}
+
+		/// <summary>
+		/// Uploads to the private catalog using the dmcatalogtoken or dmcatalogtoken environment variable as the token.
+		/// </summary>
+		/// <returns>If the upload was successful or not.</returns>
+		/// <exception cref="InvalidOperationException">Uploading failed.</exception>
+		/// <exception cref="UnauthorizedAccessException">Uploading failed due to invalid Token.</exception>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:101]:
```suggestion
		/// <exception cref="UnauthorizedAccessException">Upload failed due to invalid Token.</exception>
```

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+		public async Task<ArtifactUploadResult> UploadAsync()
+		{
+			if (String.IsNullOrWhiteSpace(keyFromEnv))
+			{
+				throw new InvalidOperationException("Uploading failed, missing token in environment variable dmcatalogtoken or dmcatalogtoken_encrypted.");
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:106]:
```suggestion
				throw new InvalidOperationException("Upload failed, missing token in environment variable dmcatalogtoken or dmcatalogtoken_encrypted.");
```

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+			}
+
+			_logger.LogDebug($"Attempting upload with Environment Variable as token for artifact: {PathToArtifact}...");
+			return await UploadAsync(keyFromEnv).ConfigureAwait(false);
+		}
+
+		/// <summary>
+		///  Attempts to find the necessary API key in Environment Variables. In order of priority:
+		///  <para>- key stored as an Environment Variable called "dmcatalogtoken". (unix/win)</para>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:115]:
```suggestion
		///  <para>- key stored as an Environment Variable called "dmcatalogtoken". (Unix/Windows)</para>
```

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		///  <para>- key configured using Skyline.DataMiner.CICD.Tools.WinEncryptedKeys called "dmcatalogtoken_encrypted" (windows only)</para>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:116]:
```suggestion
		///  <para>- key configured using Skyline.DataMiner.CICD.Tools.WinEncryptedKeys called "dmcatalogtoken_encrypted" (Windows only)</para>
```

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+						}
+					}
+				}
+				catch (InvalidOperationException)
+				{
+					// Gobble up, no key means we try the next thing.
+				}
+			}
+
+			//var config = new ConfigurationBuilder()
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:142]: remove commented out code

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+	/// <summary>
+	/// Represents all metadata for a package.
+	/// </summary>
+	public class CatalogMetaData
+	{
+		private bool artifactHadBuildNumber;
+
+		/// <summary>
+		/// The Branch/Range/Category this version belongs to. Defaults to "main"
+		/// </summary>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:19]: add <value> xml comments for all public properties

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+	/// <summary>
+	/// Represents all metadata for a package.
+	/// </summary>
+	public class CatalogMetaData
+	{
+		private bool artifactHadBuildNumber;
+
+		/// <summary>
+		/// The Branch/Range/Category this version belongs to. Defaults to "main"
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:18]:
```suggestion
		/// The Branch/Range/Category this version belongs to. Defaults to "main".
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+		/// </summary>
+		public string Branch { get; set; } = "main";
+
+		/// <summary>
+		/// The e-mail address of the Author, often the committer of a GIT Tag on the sourcecode making the package.
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:23]:
```suggestion
		/// The e-mail address of the Author, often the committer of a Git tag on the sourcecode making the package.
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+		/// </summary>
+		public string CommitterMail { get; set; }
+
+		/// <summary>
+		/// The type of content, as understood by the ArtifactUpload.
+		/// </summary>
+		public string ContentType { get; set; }
+
+		/// <summary>
+		/// An global, readable, unique identifier for this package. This is often the URI to the sourcecode.
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:33]:
```suggestion
		/// An global, readable, unique identifier for this package. This is often the URI to the source code.
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+		/// </summary>
+		public string Identifier { get; set; }
+
+		/// <summary>
+		/// The Name of the Package
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:38]:
```suggestion
		/// The name of the Package.
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+		///  The version of the package.
+		/// </summary>
+		public string Version { get; set; }
+
+		/// <summary>
+		/// Creates a partial CataLogMetaData using any information it can from the artifact itself. Check the items for null and complete.
+		/// </summary>
+		/// <param name="pathToArtifact">Path to the artifact.</param>
+		/// <returns>An instance of <see cref="CatalogMetaData"/>.></returns>
+		/// <exception cref="ArgumentNullException">Provided path should not be null</exception>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:57]:
```suggestion
		/// <exception cref="ArgumentException"><paramref name="pathToArtifact"/> is <see langword="null"/> or white space.</exception>
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+		/// <exception cref="InvalidOperationException">Expected data was not present in the Artifact.</exception>
+		public static CatalogMetaData FromArtifact(string pathToArtifact)
+		{
+			CatalogMetaData meta;
+
+			if (String.IsNullOrWhiteSpace(pathToArtifact))
+			{
+				throw new ArgumentNullException(nameof(pathToArtifact));
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:65]:
```suggestion
				throw new ArgumentException(nameof(pathToArtifact));
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+				throw new InvalidOperationException($"Invalid path to artifact. Expected a path that ends with .dmapp or .dmprotocol but received {pathToArtifact}");
+			}
+
+			return meta;
+		}
+
+		/// <summary>
+		/// Used during unit testing to assert data.
+		/// </summary>
+		/// <param name="obj"></param>
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:87]: complete

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+				   ContentType == data.ContentType;
+		}
+
+		/// <summary>
+		/// Needed to match with Equals
+		/// </summary>
+		/// <returns></returns>
+		public override int GetHashCode()
+		{
+			return HashCode.Combine(Version, Branch, Identifier, Name, ContentType, CommitterMail, ReleaseUri);
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogMetaData.cs:107]: implementing gethashcode on mutable members can cause issues.

File: CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogServiceFactory.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+		public static ICatalogService CreateWithHttp(HttpClient httpClient, ILogger logger)
+		{
+			var environment = Environment.GetEnvironmentVariable("Skyline-deploy-action-namespace");
+
+			string apiBaseUrl;
+			if (environment != null)
+			{
+				apiBaseUrl = $"https://api-{environment}.dataminer.services/{environment}";
+				logger.LogDebug("Found the \"Skyline-deploy-action-namespace\" environment variable");
+				logger.LogDebug("Setting the base url for the api to: {0}", apiBaseUrl);
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/CatalogServiceFactory.cs:29]:
```suggestion
				logger.LogDebug("Setting the base URL for the API to: {0}", apiBaseUrl);
```

File: CICD.Tools.CatalogUpload.Lib/CatalogService/HttpCatalogService.cs [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+			_logger.LogError($"The upload api returned a {response.StatusCode} response. Body: {response.Content}");
+			if (response.StatusCode is HttpStatusCode.Forbidden or HttpStatusCode.Unauthorized)
+			{
+				throw new AuthenticationException($"The upload api returned a {response.StatusCode} response. Body: {response.Content}");
+			}
+
+			throw new InvalidOperationException($"The upload api returned a {response.StatusCode} response. Body: {response.Content}");
+		}
+
+		public void Dispose()
```
Comment (PedroDebevere) [inline CICD.Tools.CatalogUpload.Lib/CatalogService/HttpCatalogService.cs:68]: implement dispose pattern

File: CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+			cancellationTokenSource = new CancellationTokenSource();
+			catalogService = service;
+			PathToArtifact = pathToArtifact;
+			TryFindEnvironmentKey();
+		}
+
+		/// <summary>
+		/// Creates an instance of <see cref="CatalogArtifact"/> using a default HttpCatalogService with a new HttpClient for communication.
+		/// It searches for an optional dmCatalogToken in the "dmcatalogtoken" or "dmcatalogtoken_encrypted" Environment Variable for authentication.
+		/// WARNING: when wishing to upload several Artifacts it's recommended to use the CatalogArtifact(string pathToArtifact, ICatalogService service, IFileSystem fileSystem, ILogger logger).
```
[the remark this answers is not in the document]
Comment (janstaelensskyline) [reply CICD.Tools.CatalogUpload.Lib/CatalogArtifact.cs:48]: Added as <remarks>

File: CICD.Tools.CatalogUpload/Program.cs [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+		private static async Task<int> Process(string pathToArtifact, string dmCatalogToken, bool isDebug)
+		{
+			return await ProcessWithRegistration(pathToArtifact, dmCatalogToken, isDebug, null, null, null, null, null);
+		}
+
+		private static async Task<int> ProcessWithRegistration(string pathToArtifact, string dmCatalogToken, bool isDebug, string registrationIdentifier, string overrideVersion, string branch, string committerMail, string releaseUri)
+		{
+			LoggerConfiguration logConfig;
+
+			if (isDebug)
```
[the remark this answers is not in the document]
Comment (janstaelensskyline) [reply CICD.Tools.CatalogUpload/Program.cs:111]: Indeed, well spotted. Adjusted.

**Pull request merged**